crates/iota-light-client/src/checkpoint.rs

```diff
@@ -0,0 +1,482 @@
+                CheckpointList::default()
+            }
+        }
+    } else {
+        CheckpointList::default()
+    };
+
+    // Try getting checkpoints from archive store if configured
+    let checkpoints_from_archive_store = if config.archive_store_config.is_some() {
+        match sync_checkpoint_list_to_latest_using_archive_store_only(config).await {
+            Ok(list) => list,
+            Err(e) => {
+                warn!("Failed to sync checkpoints from archive store: {e}");
+                CheckpointList::default()
+            }
+        }
+    } else {
+        CheckpointList::default()
+    };
+
+    // Verify we have at least some checkpoints
+    if checkpoints_from_object_store.checkpoints.is_empty()
+        && checkpoints_from_archive_store.checkpoints.is_empty()
+    {
+        bail!("Could not retrieve any checkpoints from configured sources");
+    }
+
+    let merged_checkpoints = merge_checkpoint_lists(
+        &checkpoints_from_object_store,
+        &checkpoints_from_archive_store,
+    );
+
+    // Try to sync from the full node
+    let checkpoints_list = if merged_checkpoints.is_empty() {
+        match sync_checkpoint_list_to_latest_using_full_node_only(config).await {
+            Ok(list) => list,
+            Err(e) => {
+                warn!("Failed to sync checkpoints from full node: {e}");
+                CheckpointList::default()
+            }
+        }
+    } else {
+        CheckpointList {
+            checkpoints: merged_checkpoints,
+        }
+    };
+
+    // Write the fetched checkpoint list to disk
+    write_checkpoint_list(config, &checkpoints_list)?;
+
+    Ok(checkpoints_list)
+}
+
+/// Merges two checkpoint lists, removing duplicates and ensuring the result is
+/// sorted
+fn merge_checkpoint_lists(list1: &CheckpointList, list2: &CheckpointList) -> Vec<u64> {
+    // Combine both lists into a HashSet to remove duplicates
+    let unique_checkpoints: HashSet<u64> = list1
+        .checkpoints
+        .iter()
+        .chain(list2.checkpoints.iter())
+        .copied()
+        .collect();
+
+    // Convert to sorted vector
+    let mut sorted_checkpoints: Vec<_> = unique_checkpoints.into_iter().collect();
+    sorted_checkpoints.sort();
+
+    sorted_checkpoints
+}
+
+// TODO deprecate
+/// Syncs the list of end-of-epoch checkpoints from the full node's REST, RPC
+/// and GraphQL endpoints alone.
+///
+/// No object store or archive store required, but only works with non-pruning
+/// nodes.
+async fn sync_checkpoint_list_to_latest_using_full_node_only(
+    config: &Config,
+) -> anyhow::Result<CheckpointList> {
+    info!("Syncing checkpoints from full node");
+
+    // Get the local checkpoint list, or create an empty one if it doesn't exist
+    let mut checkpoints_list = match read_checkpoint_list(config) {
+        Ok(list) => list,
+        Err(e) => {
+            info!("Could not read existing checkpoint list, starting with empty list: {e}");
```

Comment:
I would change this, otherwise the message looks weird `Could not read existing checkpoint list, starting with empty list: No such file or directory (os error 2)`
```suggestion
            info!("Could not read existing checkpoint list: {e}, starting with empty list");
```

Reply:
@Alex6323 github shows outdated, but the comment is still valid